### PR TITLE
Add manually-specified source code links to code

### LINKED ISSUE #1454: Would like to be able to link from DartDocs back to source files.
Lots of reasons this could be useful, e.g.
1.  Pointing users to where the docs were generated from so they can edit them.
2.  Pointing users to the implementations of the classes in question so they can read them.

a. came up from @inmatrix wanting to be able to encourage our users to improve our docs.
b.  Occurred to me this morning when wanting to get to the source of "TabBar" but not wanting to go have to search github when I had already found the docs:
https://docs.flutter.io/flutter/material/TabBar-class.html

Maybe the --cross-dart option already does something like this?  FYI @Hixie

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #1454 and #1892.   This adds a material icon for a source code link and calculates relative paths to insert into a link template specified by the user in a dartdoc_options.yaml file or on the command line.  Verified the alignment in mobile and normal usage.

screenshot from 2019-01-29 11-28-15
screenshot from 2019-01-29 11-29-33

### CODE REVIEW

File: lib/src/model.dart [hunk 1]
```diff
@@ -3215,6 +3216,15 @@ abstract class ModelElement extends Canonicalization
     return _documentationFrom;
   }
 
+  bool get hasSourceHref => sourceHref != '';
```
Comment (devoncarew) [inline lib/src/model.dart]: `.isNotEmpty`?

File: lib/src/model.dart [hunk 2]
```diff
@@ -3215,6 +3216,15 @@ abstract class ModelElement extends Canonicalization
...
+  String _sourceHref;
+  String get sourceHref {
+    if (_sourceHref == null) {
```
Comment (devoncarew) [inline lib/src/model.dart]: `_sourceHref ??= new SourceLinker.fromElement(this).href();` ?

File: lib/src/model.dart (hunk 1 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

File: lib/src/model.dart (hunk 2 above)
Comment (jcollins-g) [reply lib/src/model.dart]: done

**Comment by jcollins-g** [comment]
Added an example and set it up to work for dartdoc itself when it gets published.

**Pull request merged**